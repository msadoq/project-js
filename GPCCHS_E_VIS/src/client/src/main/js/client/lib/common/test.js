import { v4 } from 'uuid';
import { tmpdir } from 'os';
import _ from 'lodash';
import path from 'path';
import sinon from 'sinon';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import deepFreeze from 'deep-freeze';
import reducer from '../store/reducers/index';

const mockStore = configureMockStore([thunk]);

function getStore(initialState) {
  return createStore(
    reducer,
    initialState,
    applyMiddleware(thunk)
  );
}

const freezeMe = o => o && deepFreeze(o);

const freezeArgs = f => (...args) => {
  const frozenArgs = args.map(arg => freezeMe(arg));
  return f(...frozenArgs);
};

const testMemoization = (selector, state, ownProps) => {
  const newState = _.cloneDeep(state);
  const newOwnProps = _.cloneDeep(ownProps);
  selector.resetRecomputations();
  expect(selector.recomputations()).toBe(0);

  const result1 = selector(newState, newOwnProps);
  expect(selector.recomputations()).toBe(1);

  const result2 = selector(newState, newOwnProps);
  expect(selector.recomputations()).toBe(1);
  expect(result1).toEqual(result2);
};

const testPayloads = [];
const testHandler = (...args) => {
  _.each(args, (arg) => {
    testPayloads.push(arg);
  });
};

module.exports = {
  sinon, // TODO: USELESS: prefer directly import sinon
  getStore, // TODO: use mockStore instead
  mockStore, // thunk testing
  freezeMe, // reducers testing
  freezeArgs, // reducers testing
  testMemoization, // reselect testing
  isV4: (id = '') => id.length === v4().length, // TODO: replace by a jest 'toBeV4' assertion
  getTmpPath: (...args) => path.resolve(tmpdir(), 'vima-tests', ...args), // documentManager testing
  testHandler, // used by serverProcess controllers tests
};
