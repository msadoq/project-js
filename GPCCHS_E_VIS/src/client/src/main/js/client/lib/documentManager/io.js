import { writeFile } from 'fs';
import startsWith from 'lodash/fp/startsWith';

import {
  readJsonFromAbsPath, readJsonFromRelativePath, readJsonFromFmdPath,
} from '../common/fs';

const readJsonFromOId = fmdApi => (oId, callback) => {
  fmdApi.resolveDocument(oId, (err, path, properties) => {
    if (err) {
      return callback(err);
    }
    readJsonFromFmdPath(path, (error, json) => callback(error, json, properties));
  });
};

export const readDocument = fmdApi => (folder, relativePath, oId, absolutePath, callback) => {
  if (absolutePath) {
    return readJsonFromAbsPath(absolutePath, callback);
  }
  if (oId) {
    return readJsonFromOId(fmdApi)(oId, callback);
  }
  if (folder && !startsWith('/', relativePath)) {
    return readJsonFromRelativePath(folder, relativePath, callback);
  }
  return readJsonFromFmdPath(relativePath, callback);
};

export const writeDocument = fmdApi => (path, json, callback) => {
  const spaces = 2; // beautify json with 2 spaces indentations
  const data = JSON.stringify(json, null, spaces);
  if (!startsWith('/', path)) {
    return callback(new Error('path should be absolute'));
  }
  if (fmdApi.isInFmd(path)) {
    return fmdApi.createDocument(path, json.type, (err, oid) => {
      if (err) {
        return callback(err);
      }
      return writeFile(path, data, (errWriting) => {
        if (errWriting) {
          return callback(errWriting);
        }
        callback(null, oid);
      });
    });
  }
  return writeFile(path, data, callback);
};
