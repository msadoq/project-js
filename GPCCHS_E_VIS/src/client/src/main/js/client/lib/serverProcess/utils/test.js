const _each = require('lodash/each');
const path = require('path');
// eslint-disable-next-line import/no-extraneous-dependencies, "DV6 TBC_CNES Only used by UT files"
const chai = require('chai');
// eslint-disable-next-line import/no-extraneous-dependencies, "DV6 TBC_CNES Only used by UT files"
const properties = require('chai-properties');
const database = require('../models/loki');

const registerDc = require('common/protobuf/adapters/dc');
const registerLpisis = require('common/protobuf/adapters/lpisis');

registerDc(path.resolve(__dirname, '../../..', 'node_modules/common/protobuf/proto/dc')); // Temporary fix for packaging
registerLpisis(path.resolve(__dirname, '../../..', 'node_modules/common/protobuf/proto/lpisis')); // Temporary fix for packaging

chai.use(properties);

let testHandlerArgs = [];

const testHandler = (...args) => {
  _each(args, (arg) => {
    testHandlerArgs.push(arg);
  });
};

// URL used into e2e tests for WS connection with HSS
const e2eUrl = () => `${process.env.E2E_URL}:${process.env.SERVER_PORT}?identity=main`;

module.exports = {
  should: expect(chai)(),
  database,
  testHandler,
  getTestHandlerArgs: () => testHandlerArgs,
  resetTestHandlerArgs: () => { testHandlerArgs = []; },
  e2eUrl,
};
