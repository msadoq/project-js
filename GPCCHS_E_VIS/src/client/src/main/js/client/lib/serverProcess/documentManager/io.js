// ====================================================================
// HISTORY
// VERSION : 1.1.2 : FA : ISIS-FT-1964 : 21/07/2017 : Move documentManager in serverProcess .
// VERSION : 1.1.2 : DM : #6700 : 03/08/2017 : Merge branch 'dev' into dbrugne-data
// END-HISTORY
// ====================================================================

import _ from 'lodash/fp';
import fs from 'fs';

import { read, parse } from 'common/fs';

import { MIME_TYPES } from 'constants';
import * as fmd from './fmd';

import resolvePath from './pathResolver';

export const readDocument = (
  { pageFolder, workspaceFolder, viewFolder, documentFolder, path, oId, absolutePath }, cb
) => {
  const folder = viewFolder || pageFolder || workspaceFolder || documentFolder;
  const relativePath = path;
  resolvePath({ folder, relativePath, oId, absolutePath }, (err, { resolvedPath, properties }) => {
    if (err) {
      return cb(err, {});
    }
    return read(resolvedPath, (readErr, content) => {
      if (readErr) {
        return cb(readErr);
      }
      return parse(content, (parseErr, parsedJson) => {
        if (parseErr) {
          return cb(parseErr);
        }
        return cb(null, parsedJson, properties, resolvedPath);
      });
    });
  });
};

export const readDocumentType = (docInfo, cb) => {
  readDocument(docInfo, (err, doc) => {
    if (err) {
      return cb(err);
    }
    if (!_.has('type', doc)) {
      return cb(new Error('The given document has no type'));
    }
    return cb(null, doc.type);
  });
};

export const writeDocument = (path, json, cb) => {
  const spaces = 2; // beautify json with 2 spaces indentations
  const data = JSON.stringify(json, null, spaces);
  if (!_.startsWith('/', path)) {
    return cb(new Error('path should be absolute'));
  }
  const writeFile = oid => fs.writeFile(path, data, (errWriting) => {
    if (errWriting) {
      return cb(errWriting);
    }
    return cb(null, oid);
  });
  if (fmd.isInFmd(path)) {
    return fmd.createDocument(path, json.type, (err, oid) => {
      if (err) {
        return cb(new Error(`FMD createDocument : ${err}`));
      }
      return writeFile(oid);
    });
  }
  return writeFile();
};

export const exportData = (content, contentType, path, callback) => {
  if (!_.startsWith('/', path)) {
    return callback(new Error('path should be absolute'));
  }
  const encoding = (type) => {
    switch (type) {
      case MIME_TYPES.PortableNetworkGraphics: {
        return 'base64';
      }
      case MIME_TYPES.CommaSeparatedValues: {
        return 'utf8';
      }
      default: {
        return 'utf8';
      }
    }
  };
  const writeFile = oid => fs.writeFile(path, content, encoding(contentType), (errWriting) => {
    if (errWriting) {
      return callback(errWriting);
    }
    return callback(null, oid);
  });
  if (fmd.isInFmd(path)) {
    return fmd.createDocument(path, contentType, (err, oid) => {
      if (err) {
        return callback(new Error(`FMD createDocument : ${err}`));
      }
      return writeFile(oid);
    });
  }
  return writeFile();
};
