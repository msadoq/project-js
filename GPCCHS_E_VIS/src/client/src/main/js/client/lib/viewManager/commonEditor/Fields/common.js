import { get } from 'common/configurationManager';

/**
 * @param list
 * @param addWildcard
 */
// eslint-disable-next-line import/prefer-default-export
export const computeOptions = (list, addWildcard = false) => {
  let options =
    list && Array.isArray(list)
      ? list.map(d =>
        ({
          label: d.name,
          value: d.name,
        })
      )
      : []
  ;
  if (addWildcard) {
    const wildcardCharacter = get('WILDCARD_CHARACTER');
    if (!options.find(s => s.value === wildcardCharacter)) {
      options = options.concat({
        label: wildcardCharacter,
        value: wildcardCharacter,
      });
    }
  }

  return options;
};
