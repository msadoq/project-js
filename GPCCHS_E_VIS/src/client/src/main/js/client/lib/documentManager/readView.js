// import _ from 'lodash/fp';
import async from 'async';
import { v4 } from 'uuid';

import { readDocument } from './io';
import validation from './validation';
import {
  isViewTypeSupported,
  getSchema,
  getViewModule,
} from '../viewManager';

const simpleReadView = async.reflect((viewInfo, cb) => {
  readDocument(viewInfo, (err, viewContent, properties, viewPath) => {
    if (err) {
      return cb(err);
    }
    if (!isViewTypeSupported(viewContent.type)) {
      return cb(new Error(`Unsupported view type '${viewContent.type}'`), viewInfo);
    }

    const schema = getSchema(viewContent.type);

    const validationError = validation(viewContent.type, viewContent, schema);
    if (validationError) {
      return cb(validationError);
    }

    const uuid = viewInfo.uuid || v4();
    const view =
      getViewModule(viewContent.type)
        .prepareViewForStore({ ...viewContent, uuid });

    return cb(null, {
      ...viewInfo,
      ...view,
      isModified: false,
      path: viewInfo.path,
      oId: viewInfo.oId,
      absolutePath: viewPath,
    });
  });
});

export default {
  simpleReadView,
};
