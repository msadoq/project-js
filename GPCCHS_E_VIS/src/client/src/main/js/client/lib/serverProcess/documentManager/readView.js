// ====================================================================
// HISTORY
// VERSION : 1.1.2 : FA : ISIS-FT-1964 : 21/07/2017 : Move documentManager in serverProcess .
// VERSION : 1.1.2 : DM : #6700 : 03/08/2017 : Merge branch 'dev' into dbrugne-data
// VERSION : 2.0.0 : DM : #5806 : 06/12/2017 : Change all relative imports .
// END-HISTORY
// ====================================================================

// import _ from 'lodash/fp';
import async from 'async';
import { v4 } from 'uuid';

import {
  isViewTypeSupported,
  getSchema,
  getViewModule,
} from 'viewManager';
import { readDocument } from './io';
import validation from './validation';

const simpleReadView = async.reflect((viewInfo, cb) => {
  readDocument(viewInfo, (err, viewContent, properties, viewPath) => {
    if (err) {
      return cb(err);
    }
    if (!isViewTypeSupported(viewContent.type)) {
      return cb(new Error(`Unsupported view type '${viewContent.type}'`), viewInfo);
    }

    const schema = getSchema(viewContent.type);

    const validationError = validation(viewContent.type, viewContent, schema);
    if (validationError) {
      return cb(validationError);
    }

    const uuid = viewInfo.uuid || v4();
    const view = getViewModule(viewContent.type).prepareViewForStore({ ...viewContent, uuid });

    return cb(null, {
      isModified: false,
      ...viewInfo,
      ...view,
      path: viewInfo.path,
      oId: viewInfo.oId,
      absolutePath: viewPath,
    });
  });
});

export default {
  simpleReadView,
};
