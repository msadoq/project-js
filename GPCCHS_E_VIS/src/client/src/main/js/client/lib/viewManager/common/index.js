import _unset from 'lodash/unset';
import _memoize from 'lodash/memoize';
import { SIGNIFICANT_VALIDITY_STATE_VALUE } from 'constants';
import { SDB_VALUE_OPTION, TIME_BASED_DATA_OPTION } from '../commonEditor/Fields/DataTypeField';

export const buildFormula = (catalog, catalogItem, comObject, comObjectField) => {
  let result = '';
  if (catalog) {
    result += catalog;
  }

  if (catalog && catalogItem) {
    result += `.${catalogItem}`;
  }

  if (catalog && catalogItem && comObject) {
    result += `<${comObject}>`;
  }

  if (catalog && catalogItem && comObject && comObjectField) {
    result += `.${comObjectField}`;
  }

  return result;
};

export const buildFormulaForAutocomplete = (catalog, catalogItem, comObject, comObjectField) => {
  let result = '';
  if (catalog) {
    result += catalog;

    if (catalogItem) {
      result += `.${catalogItem}`;

      if (comObject) {
        result += `<${comObject}>`;

        if (comObject.indexOf('ReportingParameter') > -1) {
          result += '.convertedValue';
          return result;
        } else if (comObject.indexOf('DecommutedPacket') > -1) {
          return result;
        }

        if (comObjectField) {
          result += `.${comObjectField}`;
        }

        return result;
      }
    }
  }
  return result;
};

/**
 * Handles entry point editor form submission
 * @param values
 * @param updateEntryPoint callback to update the entry point
 * @param viewId
 */
export function handleSubmit(values, updateEntryPoint, viewId) {
  const { dataType } = values.connectedData;
  if (dataType !== SDB_VALUE_OPTION.value) {
    _unset(values.connectedData, 'path');
    _unset(values.connectedData, 'displayMode');
  }
  if (dataType !== TIME_BASED_DATA_OPTION.value) {
    _unset(values.connectedData, 'catalogItem');
    _unset(values.connectedData, 'comObject');
    _unset(values.connectedData, 'comObjectField');
    _unset(values.connectedData, 'provider');
    _unset(values.connectedData, 'refTimestamp');
  }
  const { catalog, catalogItem, comObject, comObjectField } = values.connectedData;
  const formula = buildFormula(catalog, catalogItem, comObject, comObjectField);
  updateEntryPoint(viewId, values.id, {
    ...values,
    connectedData: {
      ...values.connectedData,
      formula,
    },
  });
}

/**
 * given an array of required fields, returns the validate function
 * @param requiredFields
 * @returns a redux-form validate function
 */
const innerValidateRequiredFields = requiredFields => (values = {}) => {
  const errors = {};

  requiredFields.forEach((field) => {
    if (!values[field]) {
      errors[field] = 'Required';
    }
  });
  return errors;
};

export const validateRequiredFields = (requiredFields, values) => (
  values
    ? innerValidateRequiredFields(requiredFields)(values) // executes the validation on values
    : innerValidateRequiredFields(requiredFields) // returns a validation function
);

/**
 * given a validity state, returns true if value is "VALID(2)" or false otherwise
 * @type {Function}
 * @returns boolean
 */
export const memoizeIsSignificantValue = _memoize(
  validityState => validityState === SIGNIFICANT_VALIDITY_STATE_VALUE
);
