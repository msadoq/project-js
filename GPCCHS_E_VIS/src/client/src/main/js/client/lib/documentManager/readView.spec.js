import { should, sinon } from '../common/test';
import { simpleReadView } from './readView';
import * as io from './io';

const stubReadDocument = (doc = {}, type = 'TextView') => sinon.stub(io, 'readDocument').callsFake((viewInfo, cb) => {
  if (viewInfo.withError) {
    return cb(new Error('Fake Error'));
  }
  const content = {
    type,
    ...doc,
  };
  return cb(null, content, null, '/a/fake/absolute/path');
});

describe('documentManager:readView', () => {
  let stub;
  afterEach(() => {
    stub.restore();
  });

  it('creates a simple view', (done) => {
    stub = stubReadDocument({ entryPoints: [], content: '' }, 'TextView');
    simpleReadView({ uuid: 'fake-uuid', path: '/a/path', oId: 'an oid' }, (err, content) => {
      content.should.be.eql({
        value: {
          uuid: 'fake-uuid',
          path: '/a/path',
          oId: 'an oid',
          type: 'TextView',
          defaultRatio: { length: 5, width: 5 },
          links: [],
          title: 'New Text View',
          configuration: { content: '', entryPoints: [] },
          isModified: false,
          absolutePath: '/a/fake/absolute/path',
        },
      });
      done();
    });
  });

  it('gives an error when view type is not supported', (done) => {
    stub = stubReadDocument({ entryPoints: [], content: '' }, 'FakeView');
    simpleReadView({ uuid: 'fake-uuid', path: '/a/path', oId: 'an oid' }, (err, content) => {
      should.not.exist(err);
      content.error.should.be.an('error');
      done();
    });
  });
  it('gives an error when readDocument failed', (done) => {
    stub = stubReadDocument({ entryPoints: [], content: '' }, 'TextView');
    simpleReadView({ withError: true }, (err, content) => {
      should.not.exist(err);
      content.error.should.be.an('error');
      done();
    });
  });
});
