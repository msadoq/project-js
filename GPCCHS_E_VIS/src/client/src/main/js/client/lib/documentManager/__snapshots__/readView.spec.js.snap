// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentManager:readView creates a simple view using readed document 1`] = `
Object {
  "value": Object {
    "absolutePath": "/a/fake/absolute/path",
    "configuration": Object {
      "content": "",
      "entryPoints": Array [],
    },
    "defaultRatio": Object {
      "length": 5,
      "width": 5,
    },
    "isModified": false,
    "links": Array [],
    "oId": "an oid",
    "path": "/a/path",
    "title": "New Text View",
    "type": "TextView",
    "uuid": "fake-uuid",
  },
}
`;

exports[`documentManager:readView creates a simple view with random uuid 1`] = `
Object {
  "value": Object {
    "absolutePath": "/a/fake/absolute/path",
    "configuration": Object {
      "content": "",
      "entryPoints": Array [],
    },
    "defaultRatio": Object {
      "length": 5,
      "width": 5,
    },
    "isModified": false,
    "links": Array [],
    "oId": "an oid",
    "path": "/a/path",
    "title": "New Text View",
    "type": "TextView",
  },
}
`;
